notebooks/02-ingestion/05-partitioning-data.ipynb

```diff
@@ -0,0 +1,717 @@
+   ]
+  },
+  {
+   "attachments": {},
+   "cell_type": "markdown",
+   "id": "8799d234-c2f3-4277-9661-8931b94d58ff",
+   "metadata": {},
+   "source": [
+    "The [segment view for this table](http://localhost:8888/unified-console.html#segments/datasource~flights_year_IATA) shows 12 segments that are still very close to our ideal segment size, but now they have some new metadata based on the clustering columns in the `Shard Spec` column. It shows which range of values of the IATA_CODE_Reporting_Airline column are available in the segment. This is very useful for pruning when filtering on that column because Druid can prune to 1 or 2 segment files when looking for a single airline.\n",
+    "\n",
+    "![](assets/segments-iata.png)\n",
+    "\n",
+    "## Query Performance\n",
+    "Now that the data has been loaded using a few different partitioning strategies, you can use a test query to determine which one is the best.\n",
+    "In order to control the experiment we will turn cacheing off for the results of the query and measure the same query with each table.\n",
+    "\n",
+    "### The Query\n",
+    "It is well known that flight traffic follows a weekly pattern as well as a seasonal pattern. Given that we only have a month, let's take a look at the maximum and average delay for a single airline by day of the week so that we can see the pattern and how it affects delays. We'll only consider the departure delay for now and only use delays greater than 2 minutes as \"delayed\".\n",
+    "\n",
+    "One caveat about measuring performance on a laptop. The results obtained here will not necessarily be the best when the data grows and is distributed on a multi-node cluster. The intent here is to provide a way of thinking about optimizing the data. In a real cluster, it is likely that the results will vary and a different partitioning and clustering strategy will provide better results.\n",
+    "\n",
+    "The following cell defines a function we can use to measure performance:"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "77a71e6a-73a2-4bd7-9e8d-fd7cb966da65",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "from datetime import datetime\n",
+    "from statistics import mean \n",
+    "\n",
+    "def measure_query( sql: str, iterations: int ):\n",
+    "    req = sql_client.sql_request(sql)\n",
+    "    req.add_context(\"populateCache\", \"false\")  # run without cacheing results to get a real sense of performance\n",
+    "    req.add_context(\"useCache\", \"false\")  # do not use cached results\n",
+    "    stats = []\n",
+    "    while (iterations>0):\n",
+    "      start = datetime.now()\n",
+    "      sql_client.sql(req)\n",
+    "      end = datetime.now()\n",
+    "      stats.append( (end - start).total_seconds() * 10**3 ) # add run time in milliseconds\n",
+    "      iterations -=1\n",
+    "    return f\"Results = avg:{mean(stats)} ms   min:{min(stats)} ms  max:{max(stats)} ms\""
```

Comment:
Maybe add a bit of commentary here to say what the function does.

Reply:
added comments in the code, above the function that describe what it does.
```
# The measure_query function runs the specified sql query mutliple times 
# and measures the duration of each run.
# The output is a string with the results the mean, minimum and maximum durations measured.
```